### ISSUE TITLE
Multiple versions

### ISSUE THREAD

**Issue by 1Computer1** [discordjs/guide#121]
Support for multiple versions would be nice. You could have one version for stable and one for master for example. The user could then choose with a dropdown menu.

**Comment by Danktuary** [comment]
I'm a bit stuck on how to go about something like this. I had something in mind for this a while ago but never went through with it because we didn't really need it.

My idea was to create a Vue component that'd look something along the lines of this:
```vue
<!-- unsure of the component name, but just as an example -->
<branch-specific branch="stable">
	Here's some code that would only work on the *stable* branch.
</branch-specific>

<branch-specific branch="master">
	Here's some code that would only work on the *master* branch.
</branch-specific>

<!-- In case there's another branch other than stable and master -->
<branch-specific :branch="['stable', 'master']">
	Here's some code that would work on both the *stable* and *master* branches, but not any other branch.
</branch-specific>
```

Because 90% of the majority of pages and their contents will stay the same. In the cases of pages that only apply to one branch and not the other, we can do something like this.

```vue
<branch except="master">
	<p class="warning">This page only applies to the *master* branch. Click [here](#) to switch to it.</p>
</branch>
```

At the moment, I think this is the most decent approach, so that we don't have to have multiple files for the same page, just on different branches. You'd only need to change one file when makes edits, and we could easily update the contents of the page with Vue without having to redirect the user to another page, refresh the page, or do anything else complex.

However, if anybody else has some ideas, I'll gladly hear them out. 😄

**Comment by iCrawl** [comment]
+1

There is absolutely no reason to duplicate pages or "switch" sites like we do on the docs site.
I think this is a great middleground to hit off on both the new "breaking" stuff and the old stable (which I'm sure quite a lot of people still use for the sake of stability).
There also shouldn't be many topics that require a "master only" tag.

Overall this should also make it easier for people to update from one version to another if the need ever occurs. Thus further reducing the thousands of articles and tags about "change X to Y on master".

**Comment by NbOpposite** [comment]
I wonder, wouldn't a version-specific component make more sense?
Like, you specify semvers that this applies for the current topic rather than a git branch, as at some point, master will become version 12, and then, when we start work on the next major release, the docs with master-branch set will suddenly not apply to the master, but rather version 12?

**Comment by Danktuary** [comment]
Good point, @NbOpposite. The component example I gave would stay roughly the same, but the dropdown menu could look something like this:

![Version chooser dropdown menu](https://i.imgur.com/ZOBkSO4.png)

And then the components could be used as:

```vue
<branch-specific version="11">
	Some content that only shows for the 11.x branches.
</branch-specific>
```

The dropdown item values would determine what branch to change to, while the actual display is still user-friendly and will reflect the current branch names.

I'm not sure how far the semver tree we wanna go, though. I think just `major.minor` would be best, because there is some stuff changed in 11.4 that isn't available in 11.3 (for example), and it wouldn't add much overhead complexity to the component logic. I'll mess around with the concept sooner or later and see how it goes. 👍

**Comment by NbOpposite** [comment]
Even better would be to be able to set a range of versions something applies to btw:

Stuff that applies to both v11 and v12 could have `version="11.3-12.x"`

**Comment by Danktuary** [comment]
Quick note: I'll probably wait until we convert to VuePress to add this in. While I do have something that works right now, I still have to design and build the actual HTML bit. If I do that now, I'll have to rewrite it once we convert either way.

Once we start using VuePress with Webpack, I can get rid of my code and take advantage of the [semver package](https://www.npmjs.com/package/semver), which would allow me to make a better component without having to worry about handling the logic myself.

**Comment by almostSouji** [comment]
Branch selection has landed on master as of 01fa3eb51b983504f8dff26ea1167d41a1b974a9.
A quick guide for its usage can be found [here (CONTRIBUTING.md)](https://github.com/discordjs/guide/blob/master/CONTRIBUTING.md#branch-specific-content)
